File: Algorithms/Python/0034_Find_First_and_Last_Position_of_Element_in_Sorted_Array/Find_First_and_Last_Position_of_Element_in_Sorted_Array_Solution_1.py

# Space: O(1)
# Time: O(logn)

class Solution:
    def searchRange(self, nums, target):
        def find_first(nums, target):
            left = 0
            right = len(nums) - 1

            while left + 1 < right:
                mid = (left + right) // 2
                if nums[mid] >= target:
                    right = mid
                else:
                    left = mid + 1
            if nums[left] == target:
                return left
            if nums[right] == target:
                return right
            return -1

        def find_last(nums, target):
            left = 0
            right = len(nums) - 1

            while left + 1 < right:
                mid = (left + right) // 2
                if nums[mid] <= target:
                    left = mid
                else:
                    right = mid - 1
            if nums[right] == target:
                return right
            if nums[left] == target:
                return left
            return -1

        if len(nums) == 0: return [-1, -1]
        return [find_first(nums, target), find_last(nums, target)]

File: Algorithms/0034_Find_First_and_Last_Position_of_Element_in_Sorted_Array/Java/FindFirstAndLastPositionOfElementInSortedArraySolution1.java

// Space: O(1)
// Time: O(logn)

class Solution {
    public int[] searchRange(int[] nums, int target) {
        int length = nums.length;
        if (length == 0) return new int[]{-1, -1};
        if (length == 1) {
            if (nums[0] == target) return new int[]{0, 0};
            return new int[]{-1, -1};
        }
        int[] res = new int[2];
        int left = 0, right = length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                right = mid - 1;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            }
        }
        if (left < length && nums[left] == target) {
            res[0] = left;
        } else {
            return new int[]{-1, -1};
        }
        left = 0;
        right = length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                left = mid + 1;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            }
        }
        res[1] = right;
        return res;
    }
}




